docs/site/REST-middleware-sequence.md

```diff
@@ -0,0 +1,846 @@
+
+    - Reject the response by throwing an error
+    - Transform the response to a different value
+    - Return the response to upstream middleware
+
+3.  Catch the error thrown from `await next()`. If the `catch` block does not
+    exist, the error will be bubbled up to upstream middleware.
+
+![cascading-middleware](imgs/cascading-middleware.png)
+
+Default sequence executes these groups of middleware in order:
+
+- `cors`: Enforces `CORS`
+- `openApiSpec`: Serves OpenAPI specs
+- `findRoute`: Finds the appropriate controller method, swagger spec and args
+  for invocation
+- `parseParams`: Parses HTTP request to get API argument list
+- `invokeMethod`: Invokes the API which is defined in the Application controller
+  method
+
+In front of the groups above, we have a special middleware called
+`SendResponse`, which first invokes downstream middleware to get a result and
+handles the result or error respectively.
+
+- Writes the result from API into the HTTP response (if the HTTP response has
+  not been produced yet by the middleware chain.
+- Catches error logs it using 'logError' if any of the above steps in the
+  sequence fails with an error.
+
+## Migrate from legacy sequence
+
+The `Sequence` generated before `@loopback/rest@6.0.0` comes with hard-coded
+actions as follows:
+
+```ts
+import {inject} from '@loopback/core';
+import {
+  FindRoute,
+  InvokeMethod,
+  InvokeMiddleware,
+  ParseParams,
+  Reject,
+  RequestContext,
+  RestBindings,
+  Send,
+  SequenceHandler,
+} from '@loopback/rest';
+
+const SequenceActions = RestBindings.SequenceActions;
+
+export class MySequence implements SequenceHandler {
+  /**
+   * Optional invoker for registered middleware in a chain.
+   * To be injected via SequenceActions.INVOKE_MIDDLEWARE.
+   */
+  @inject(SequenceActions.INVOKE_MIDDLEWARE, {optional: true})
+  protected invokeMiddleware: InvokeMiddleware = () => false;
+
+  constructor(
+    @inject(SequenceActions.FIND_ROUTE) protected findRoute: FindRoute,
+    @inject(SequenceActions.PARSE_PARAMS) protected parseParams: ParseParams,
+    @inject(SequenceActions.INVOKE_METHOD) protected invoke: InvokeMethod,
+    @inject(SequenceActions.SEND) public send: Send,
+    @inject(SequenceActions.REJECT) public reject: Reject,
+  ) {}
+
+  async handle(context: RequestContext) {
+    try {
+      const {request, response} = context;
+      const finished = await this.invokeMiddleware(context);
+      if (finished) return;
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+      this.send(response, result);
+    } catch (err) {
+      this.reject(context, err);
+    }
+  }
+}
+```
+
+The legacy `Sequence` is now deprecated but it should continue to work without
+any changes.
+
+If you have never customized `src/sequence.ts` generated by `lb4 app`, or you
+have only modified the code by adding `authenticate` action per our docs, you
```

Comment:
can refer to https://loopback.io/doc/en/lb4/Authentication-component-action.html